### PR TITLE
Control Allocation: fixes in yaw saturation detection for vehicles with tilt-for-yaw

### PR DESCRIPTION
**Pull request by sfuhrer**
### Solved Problem
VTOL Tiltrotors require a special yaw saturation detection. Simply looking at whether there is unallocated yaw torque leads to almost constant saturation reporting, as often one of the tilts is saturated (they usually can't tilt backwards). I patched this previously in https://github.com/PX4/PX4-Autopilot/pull/20518, where I simply declared the yaw saturated if the torque setpoint was saturated. Given that there is also a  collective tilt this assumption is quite rough. 

While flying a tricopter with sideways tiltable tail I further noticed that there the reported unallocated yaw was never 0, while the tilt servo clearly never reached saturation:
image
I didn't investigate further why the unallocated torque calculation is so wrong, but given that MC tilt vehicles would anyway have the same issues as VTOL tiltrotor if they have tilts that only tilt in one direction, it makes sense to me to customize the saturation handling anyway also for them.

### Solution
Customize yaw saturation detection for VTOL tiltrotor and MC tilts. Taking the sign of the yaw effectiveness of the corresponding tilt into account, check if the tilt is at the min or max value and set the positive or negative saturation flag if all the tilts are saturated in this direction. 

### Changelog Entry
For release notes:
```
Bugfix: fix yaw saturation logic for tilting motors (tiltrotor VTOL and MC with tilts)
```

### Alternatives
Extend it to have it generic for all axis with tilt actuation. I wonder though if we also should first remove the tilts from the effectiveness matrix and do custom allocation of them (something similar to what is done in the Helicopter allocation).

### Test coverage
SITL tested (tiltrotor).

### CODE REVIEW

File: src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessTiltrotorVTOL.cpp [hunk 1]
```diff
@@ -137,12 +137,39 @@ void ActuatorEffectivenessTiltrotorVTOL::updateSetpoint(const matrix::Vector<flo
...
+
 			for (int i = 0; i < _tilts.count(); ++i) {
 				if (_tilts.config(i).tilt_direction == ActuatorEffectivenessTilts::TiltDirection::TowardsFront) {
 					actuator_sp(i + _first_tilt_idx) += control_collective_tilt;
 				}
+
+				// custom yaw saturation logic: only declare yaw saturated if all tilts are at the negative or positive yawing limit
+				if (_tilts.getYawTorqueOfTilt(i) > FLT_EPSILON) {
+
+					if (yaw_saturated_positive && actuator_sp(i + _first_tilt_idx) < actuator_max(i + _first_tilt_idx)) {
```
Comment (bkueng) [inline src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessTiltrotorVTOL.cpp:151]: I would check with a margin here as well:
`actuator_sp(i + _first_tilt_idx) < actuator_max(i + _first_tilt_idx) - FLT_EPSILON`

File: src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessTiltrotorVTOL.cpp [hunk 2]
```diff
@@ -137,12 +137,39 @@ void ActuatorEffectivenessTiltrotorVTOL::updateSetpoint(const matrix::Vector<flo
...
+						yaw_saturated_positive = false;
+					}
+
+					if (yaw_saturated_negative && actuator_sp(i + _first_tilt_idx) > actuator_min(i + _first_tilt_idx)) {
+						yaw_saturated_negative = false;
+					}
+
+				} else if (_tilts.getYawTorqueOfTilt(i) < FLT_EPSILON) {
```
Comment (bkueng) [inline src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessTiltrotorVTOL.cpp:159]: `_tilts.getYawTorqueOfTilt(i) < -FLT_EPSILON`

**Review by bkueng — COMMENTED** [review]
Makes sense.

File: src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessTiltrotorVTOL.cpp (hunk 1 above)
Comment (sfuhrer) [reply src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessTiltrotorVTOL.cpp:151]: agree, done

File: src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessTiltrotorVTOL.cpp (hunk 2 above)
Comment (sfuhrer) [reply src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessTiltrotorVTOL.cpp]: goot catch, changed it

**Review by bkueng — APPROVED** [review]

**Pull request merged**